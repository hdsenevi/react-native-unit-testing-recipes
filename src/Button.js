import React, { Component } from 'react';
import { Text, TouchableOpacity, Platform, Linking } from 'react-native';

// 1. Changed to a class based component
class Button extends Component {
    constructor(props) {
        super(props);
    }

    // 2. Custom function called onPress TouchableOpacity
    onPressHandler = () => {
        const { onPress, url } = this.props
        if (url) {
            Linking.openURL(url)
        }
        onPress()
    }

    render(){
        const { buttonStyle, textStyle } = styles;
        const { label, primary } = this.props;
        
        // 3. Custom platform specific rendering
        const height = Platform.OS === 'ios' ? 45 : 65;

        // 4. Change color of button depending on 'primary' prop
        const newButtonStyle = primary ? 
            [buttonStyle, { height }] : 
            [buttonStyle, {height, backgroundColor: '#f34541', borderBottomColor: '#a43532'}];

        // 5. Platform specific font styling
        const fontSize = Platform.OS === 'ios' ? 16 : 20;

        return (
            <TouchableOpacity onPress={this.onPressHandler} style={newButtonStyle}>
                <Text style={[textStyle, { fontSize }]}>
                    {label}
                </Text>
            </TouchableOpacity>
        );
    }
};

Button.defaultProps = {
    primary: true,
}

const styles = {
    textStyle: {
        alignSelf: 'center',
        color: '#fff',
        fontWeight: '600',
    },
    buttonStyle: {
        alignSelf: 'stretch',
        justifyContent: 'center',
        backgroundColor: '#38ba7d',
        borderBottomWidth: 6,
        borderBottomColor: '#1e6343',
        borderWidth: 1,
        marginLeft: 15,
        marginRight: 15
    }
};

export default Button;
