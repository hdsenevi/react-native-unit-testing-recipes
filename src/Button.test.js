import React from 'react';
import {shallow} from 'enzyme';
import Button from './Button';

let mockOpenURL = jest.fn();
jest.mock('Linking', () => ({
    openURL: mockOpenURL,
}));

describe('Button', () => {
    describe('Rendering', () => {
        it('should match to snapshot - Primary', () => {
            const component = shallow(<Button label="test label" primary />)
            expect(component).toMatchSnapshot("Primary button snapshot")
        });
        it('should match to snapshot - Secondary', () => {
            const component = shallow(<Button label="test label" primary={false} />)
            expect(component).toMatchSnapshot("Secondary button snapshot")
        });
    });

    describe('Interaction', () => {
        describe('onPressHandler', () => {
            const mockOnPress = jest.fn();
            let instance;

            beforeEach(() => {
                instance = shallow(<Button 
                    label="test label" 
                    onPress={mockOnPress}
                    url="https://www.test.com"
                />).instance();
                jest.clearAllMocks();
            });

            it('should call onPress', () => {
                instance.onPressHandler();
                expect(mockOnPress).toHaveBeenCalled();
                expect(mockOnPress).toHaveBeenCalledTimes(1);
            });
            
            it('should call openURL if url is provided', () => {
                instance.onPressHandler();
                expect(mockOpenURL).toHaveBeenCalled();
                expect(mockOpenURL).toHaveBeenCalledTimes(1);
                expect(mockOpenURL).toHaveBeenCalledWith("https://www.test.com");
            });

            it('should not call openURL if url is nor provided', () => {
                const innerInstance = shallow(<Button label="test label" onPress={mockOnPress} />).instance();
                innerInstance.onPressHandler();
                expect(mockOpenURL).not.toHaveBeenCalled();
            });
        });
    });
});
