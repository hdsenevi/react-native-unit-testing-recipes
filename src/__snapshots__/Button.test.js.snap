// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Rendering should match to snapshot - Primary: Primary button snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    label="test label"
    primary={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activeOpacity": 0.2,
      "children": <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
      >
        test label
      </Text>,
      "onPress": [Function],
      "style": Object {
        "alignSelf": "stretch",
        "backgroundColor": "#38ba7d",
        "borderBottomColor": "#1e6343",
        "borderBottomWidth": 6,
        "borderWidth": 1,
        "height": 45,
        "justifyContent": "center",
        "marginLeft": 15,
        "marginRight": 15,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "test label",
        "style": Object {
          "alignSelf": "center",
          "color": "#fff",
          "fontSize": 16,
          "fontWeight": "600",
        },
      },
      "ref": null,
      "rendered": "test label",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeOpacity": 0.2,
        "children": <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#fff",
              "fontSize": 16,
              "fontWeight": "600",
            }
          }
        >
          test label
        </Text>,
        "onPress": [Function],
        "style": Object {
          "alignSelf": "stretch",
          "backgroundColor": "#38ba7d",
          "borderBottomColor": "#1e6343",
          "borderBottomWidth": 6,
          "borderWidth": 1,
          "height": 45,
          "justifyContent": "center",
          "marginLeft": 15,
          "marginRight": 15,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "test label",
          "style": Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "600",
          },
        },
        "ref": null,
        "rendered": "test label",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Button Rendering should match to snapshot - Secondary: Secondary button snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    label="test label"
    primary={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activeOpacity": 0.2,
      "children": <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
      >
        test label
      </Text>,
      "onPress": [Function],
      "style": Array [
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#38ba7d",
          "borderBottomColor": "#1e6343",
          "borderBottomWidth": 6,
          "borderWidth": 1,
          "height": 45,
          "justifyContent": "center",
          "marginLeft": 15,
          "marginRight": 15,
        },
        Object {
          "backgroundColor": "#f34541",
          "borderBottomColor": "#a43532",
        },
      ],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "test label",
        "style": Object {
          "alignSelf": "center",
          "color": "#fff",
          "fontSize": 16,
          "fontWeight": "600",
        },
      },
      "ref": null,
      "rendered": "test label",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeOpacity": 0.2,
        "children": <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#fff",
              "fontSize": 16,
              "fontWeight": "600",
            }
          }
        >
          test label
        </Text>,
        "onPress": [Function],
        "style": Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#38ba7d",
            "borderBottomColor": "#1e6343",
            "borderBottomWidth": 6,
            "borderWidth": 1,
            "height": 45,
            "justifyContent": "center",
            "marginLeft": 15,
            "marginRight": 15,
          },
          Object {
            "backgroundColor": "#f34541",
            "borderBottomColor": "#a43532",
          },
        ],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "test label",
          "style": Object {
            "alignSelf": "center",
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "600",
          },
        },
        "ref": null,
        "rendered": "test label",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
